lib/FetchEnvs.js

```diff
@@ -14,235 +14,189 @@
+    const strategy = conf.overrideStrategy;
+    const {
+      name,
+      namespace = this.namespace,
+      kind = KIND_MAP.get(valueFrom),
+      key,
+      matchLabels,
+      apiVersion = 'v1',
+      type
+    } = ref;
 
-        if (data) {
-          mapData.data = data;
-        } else if (!optional) {
-          throw new Error(`envFrom.configMapRef.${name}.data not found: ${kubeError}`);
-        } else {
-          let msg = `envFrom.configMapRef.${name}.data not found: ${kubeError}`;
-          log.warn(msg);
-          this.updateRazeeLogs('warn', { controller: 'FetchEnvs', message: msg });
-        }
-      } else if (objectPath.has(element, 'secretMapRef')) {
-        let name = objectPath.get(element, 'secretMapRef.name');
-        let namespace = objectPath.get(element, 'secretMapRef.namespace', this.namespace);
-        let res;
-        try {
-          res = await this.kubeResourceMeta.request({ uri: `/api/v1/namespaces/${namespace}/secrets/${name}`, json: true });
-        } catch (e) {
-          log.warn(e);
-          kubeError = e.message;
-        }
+    const krm = await this.kubeClass.getKubeResourceMeta(
+      apiVersion,
+      kind,
+      'update'
+    );
 
-        let data = objectPath.get(res, 'data');
-        if (kubeError === undefined) kubeError = 'no data returned from processEnvFrom. make sure your data exists in the correct location and is in the expected format.';
+    const qs = labelSelectors(matchLabels);
+    const uri = krm.uri({namespace, kind, name});
+    const response = await this.api({
+      uri: uri,
+      json: true,
+      qs: qs
+    });
 
-        if (data) {
-          for (const [key, value] of Object.entries(data)) {
-            data[key] = Buffer.from(value, 'base64').toString();
-          }
-          mapData.data = data;
-        } else if (!optional) {
-          throw new Error(`envFrom.secretMapRef.${name}.data not found: ${kubeError}`);
-        } else {
-          let msg = `envFrom.secretMapRef.${name}.data not found: ${kubeError}`;
-          log.warn(msg);
-          this.updateRazeeLogs('warn', { controller: 'FetchEnvs', message: msg });
-        }
-      } else if (objectPath.has(element, 'genericMapRef')) {
-        let apiVersion = objectPath.get(element, 'genericMapRef.apiVersion');
-        let kind = objectPath.get(element, 'genericMapRef.kind');
-        let krm = await this.kubeClass.getKubeResourceMeta(apiVersion, kind, 'update');
-        let name = objectPath.get(element, 'genericMapRef.name');
-        let namespace = objectPath.get(element, 'genericMapRef.namespace', this.namespace);
-        let resource = {};
-        if (krm) {
-          try {
-            resource = await krm.get(name, namespace);
-          } catch (e) {
-            log.warn(e);
-            kubeError = e.message;
-          }
-        }
+    if (matchLabels) {
+      const output = response.items.reduce(
+        reduceItemList(ref, strategy),
+        Object.create(null)
```

Comment:
why use `Object.create(null)` instead of `{}`?

Reply:
It isn't subject to prototype shenanigans, because it doesn't have one.
Its also less likely to cause sudden v8 class transitions when adding properties dynamically.